controllers/bindings/client_factory.go

```diff
@@ -0,0 +1,321 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bindings
+
+import (
+	"context"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"time"
+
+	api "github.com/redhat-appstudio/remote-secret/api/v1beta1"
+	"github.com/redhat-appstudio/remote-secret/pkg/logs"
+	auth "k8s.io/api/authentication/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/cache"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	DefaultMaxClientCacheTTL = 30 * time.Second
+)
+
+var (
+	noAuthServiceAccountFound                         = errors.New("No service account labeled with '" + api.RemoteSecretAuthServiceAccountLabel + "' was found that could be used to authenticate deployments to targets in the local cluster")
+	onlyOneAuthServiceAccountPerNamespaceAllowed      = errors.New("there can be only one service account labeled with '" + api.RemoteSecretAuthServiceAccountLabel + "' in a namespace")
+	noKubeConfigSpecifiedForConnectionToRemoteCluster = errors.New("a secret with kubeconfig with credentials for connecting to a remote cluster is required")
+)
+
+// ClientFactory is a helper interface for the RemoteSecretReconciler that creates clients that are able to deploy to remote secret targets. The default (and only)
+// implementation is the CachingClientFactory but is hidden behind an interface so that this can be mocked out in the tests.
+type ClientFactory interface {
+	// GetClient returns a client that can be used to deploy to a target described by the targetSpec and targetStatus from a remote secret in the provided namespace
+	GetClient(ctx context.Context, currentNamespace string, targetSpec *api.RemoteSecretTarget, targetStatus *api.TargetStatus) (client.Client, error)
+	// ServiceAccountChanged signals to the client factory that the service account changed. The client factory might react by revoking the client associated with
+	// the service account from a cache, if any, etc.
+	ServiceAccountChanged(sa client.ObjectKey)
+}
+
+// LocalClusterConnectionDetails provides the client and configuration for connecting to the local cluster
+type LocalClusterConnectionDetails struct {
+	Client client.Client
+	Config *rest.Config
+}
+
+type CachingClientFactory struct {
+	// LocalCluster provides the client and configuration for connecting to the local cluster
+	LocalCluster LocalClusterConnectionDetails
+	// ClientConfigurationInitializer is given the opportunity to configure the rest configuration and client options before
+	// the client is created in the factory.
```

Comment:
I think the cost would be quite big. On every reconciliation we would be creating a (number of) clients that would potentially connect to remote clusters. Because these are "normal" kubernetes clients, they make a number of requests to the cluster to figure out what resources can be handled by that cluster (during the discovery process). This would happen every time. If there wasn't a cache, we would have to even have some logic that would figure out that two targets on the same RS can reuse the same client (if they point to the same cluster), otherwise each of them would do their own discovery requests.

In another words, caching enables us to save a whole lot of traffic to the clusters for the discovery logic.